### PR TITLE
feat: move sendpointProvider and signingRegionProvider to client metadata[WIP]

### PR DESCRIPTION
**Pull request by AllanFly120**
So that they can be generated separated to the Smithy API code generation

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/config-resolver/src/components.ts [hunk 1]
```diff
@@ -155,7 +165,7 @@ export namespace EndpointsConfig {
     /**
      * The endpoint provider to call if no endpoint is provided
      */
-    endpointProvider?: any;
+    endpointProvider?: (region: string, options?: any) => HttpEndpoint;
```
Comment (AllanFly120) [inline packages/config-resolver/src/components.ts]: Leave public interfaces' options as `any` on purpose as we want it to be private API

File: packages/config-resolver/src/components.ts [hunk 2]
```diff
@@ -92,9 +92,19 @@ export interface AWSClientRuntimeConfiguration extends RuntimeDependencies {
   signingName?: string;
 
   /**
-   * The service name with which to construct endpoints.
+   * The name of the service. By default same to endpoint prefix
```
Comment (srchase) [inline packages/config-resolver/src/components.ts]:
```suggestion
   * The name of the service. By default, used as endpoint prefix.
```

File: packages/config-resolver/src/components.ts [hunk 3]
```diff
@@ -92,9 +92,19 @@ export interface AWSClientRuntimeConfiguration extends RuntimeDependencies {
...
+
+  /**
+   * The function that provides the default endpoint
+   */
+  endpointProvider?: (region: string, options?: any) => HttpEndpoint;
+
+  /**
+   * The function that provides the default region used to create the signer
+   */
+  signingRegionProvider?: (region: string) => string;
```
Comment (mtdowling) [inline packages/config-resolver/src/components.ts]: Why is "signingRegionProvider" necessary? Why is this not part of the endpointProvider? My PR here has an example of what I'm talking about, and I think it results in a simpler system with fewer things to worry about configuring: https://github.com/aws/aws-sdk-js-v3/pull/472

**Comment by AllanFly120** [comment]
close this as #472 lands

**Pull request closed without merging**